gxutil/pm.go

```diff
@@ -775,3 +670,208 @@ func setInstallPathCache(env string, global bool, val string) {
+	addedDepCount := depQueue.AddPackageDependencies(pkg)
+	pm.ProgMeter.AddTodos(addedDepCount)
+
+	// Counter to keep track of spawned goroutines, it's not locked as it's
+	// only handled by this function. Decremented any time a message is received
+	// on the above channels, which indicates that the goroutine has finished.
+	activeGoroutines := 0
+
+	// Main loop of the function.
+	for {
+		// If there are no more dependencies to install and there aren't any
+		// goroutines running (which could potentially add new dependencies
+		// to the queue) we're finished.
+		if depQueue.Len() == 0 && activeGoroutines == 0 {
+			return nil
+		}
+
+		// Keep spawning goroutines until the allowed maximum or until
+		// there are no new dependencies to fetch at the moment.
+		for activeGoroutines < maxGoroutines && depQueue.Len() > 0 {
+
+			// Goroutine that only calls `GetPackageTo` to fetch the dependency,
+			// it either returns it or returns an error.
+			go func(dep *Dependency) {
+
+				pkgDir := filepath.Join(location, "gx", "ipfs", dep.Hash)
+				// TODO: Encapsulate in a function. Used in too many places
+				// and is part of the standard.
+
+				pm.ProgMeter.AddEntry(dep.Hash, dep.Name, "[fetch]   <ELAPSED>"+dep.Hash)
+				pkg, err := pm.GetPackageTo(dep.Hash, pkgDir)
+
+				// Either with error or with the package the goroutine ends here.
+				if err != nil {
+					fetchErrs <- fmt.Errorf("failed to fetch package: %s: %s", dep.Hash, err)
+					pm.ProgMeter.Error(dep.Hash, err.Error())
+				} else {
+					fetchedPackages <- pkg
+					pm.ProgMeter.Finish(dep.Hash)
+				}
+			}(depQueue.GetDependency())
+
+			activeGoroutines++
+		}
+
+		// Once all the possible goroutines have been spawned wait
+		// for anyone to finish and analyze (restart loop) if more
+		// goroutines can be called.
+		select {
+		case fetchedPkg := <-fetchedPackages:
+			VLog("fetched dep: %s", fetchedPkg.Name)
+			addedDepCount := depQueue.AddPackageDependencies(fetchedPkg)
+			pm.ProgMeter.AddTodos(addedDepCount)
+			activeGoroutines--
+		case err := <-fetchErrs:
+			Error("parallel fetch: %s", err)
+			return err
+			// The rest of the goroutines already running are guaranteed to have
+			// space in the channels to write to them and finish so we can end
+			// the main function here, there's not need to wait on them.
+			//
+			// TODO: If `GetPackageTo` or the `shell.Get()` function had a `Context`
+			// it would be useful to issue a `cancel()` here before returning.
+		}
+	}
+}
+
+// Call the `post-install` hook on each of the dependencies of this package
+// (direct or transitive).
+//
+// TODO: This function could also use the same parallel goroutine processing
+// structure of `fetchDependencies` but right now the `post-install` hook
+// of the only sub-tool (`gx-go rewrite`) already does a parallel processing
+// of its own, so there's little to gain here.
+func (pm *PM) dependenciesPostInstall(pkg *Package, location string) error {
```

Comment:
Can we not do this while downloading? Not a big issue but I believe it would simplify things.

Reply:
> I believe it would simplify things.

Hmm, could you expand on how do you see it as a simplification? I actually split it because it made it easy to parallelize the fetching operations without having to think about the posterior rewrite process.

Reply:
My thinking was that we could do it in the same worker. That is:

1. Download a package.
2. Run post-install immediately.

That way we don't need to traverse the tree twice.

Reply:
The other issue (I'm just remembering now, sorry) is that because you need the transitive dependencies when computing rewrites (as you we're discussing in https://github.com/whyrusleeping/gx-go/pull/38#issuecomment-396455663), the `post-install` hook needs a particular order in the DAG traversing (basically that you already fetched all the dependencies -directs and transitives- of that package before rewriting it) and that constraint invalidates the parallel fetch that's happening here (process the first dependency fetched -don't really care where it is in the DAG- and order the fetch of its dependencies) where the DAG is traversed in a somewhat stochastic manner.

(I should add this explanation to the commit message.)

Reply:
I see. Yeah, that makes sense.